src/graphics/clouds.c

```diff
@@ -0,0 +1,285 @@
+#include "clouds.h"
+
+#include "core/config.h"
+#include "core/image.h"
+#include "core/random.h"
+#include "core/speed.h"
+#include "graphics/renderer.h"
+
+#include <math.h>
+#include <string.h>
+
+#define NUM_CLOUD_ELLIPSES 30
+#define CLOUD_ALPHA_CUTOFF 32
+
+#define CLOUD_WIDTH 64
+#define CLOUD_HEIGHT 64
+
+#define CLOUD_SCALE 12
+
+#define CLOUD_MIN_CREATION_TIMEOUT 200
+#define CLOUD_MAX_CREATION_TIMEOUT 2400
+
+#define CLOUD_ROWS 4
+#define CLOUD_COLUMNS 4
+#define NUM_CLOUDS (CLOUD_ROWS * CLOUD_COLUMNS)
+
+#define CLOUD_TEXTURE_WIDTH (CLOUD_WIDTH * CLOUD_COLUMNS)
+#define CLOUD_TEXTURE_HEIGHT (CLOUD_HEIGHT * CLOUD_ROWS)
+
+#define CLOUD_SPEED 0.6
+
+typedef enum {
+    STATUS_INACTIVE,
+    STATUS_CREATED,
+    STATUS_MOVING
+} cloud_status;
+
+typedef struct {
+    int x;
+    int y;
+    int width;
+    int height;
+    int half_width;
+    int half_height;
+    int radius;
+    int squared_width;
+    int squared_height;
+    int width_times_height;
+} ellipse;
+
+typedef struct {
+    image img;
+    int x;
+    int y;
+    cloud_status status;
+    struct {
+        speed_type x;
+        speed_type y;
+    } speed;
+    float scale_x;
+    float scale_y;
+    int side;
+    int angle;
+} cloud_type;
+
+static cloud_type clouds[NUM_CLOUDS];
+static int movement_timeout;
+
+static int random_from_min_to_range(int min, int range)
+{
+    return min + random_between_from_stdlib(0, range);
+}
+
+static void position_ellipse(ellipse *e, int cloud_width, int cloud_height)
+{
+    double angle = random_fractional_from_stdlib() * M_PI * 2;
+
+    e->x = CLOUD_WIDTH / 2 + random_fractional_from_stdlib() * cloud_width * cos(angle);
+    e->y = CLOUD_HEIGHT / 2 + random_fractional_from_stdlib() * cloud_height * sin(angle);
+
+    e->width = random_from_min_to_range(CLOUD_WIDTH * 0.10, CLOUD_WIDTH * 0.10);
+    e->height = random_from_min_to_range(CLOUD_HEIGHT * 0.10, CLOUD_HEIGHT * 0.10);
+
+    e->half_width = e->width / 2;
+    e->half_height = e->height / 2;
+    e->squared_width = e->width * e->width;
+    e->squared_height = e->height * e->height;
+
+    e->width_times_height = e->squared_width * e->squared_height;
+    e->radius = e->half_width > e->half_height ? e->half_width : e->half_height;
+}
+
+static int ellipse_is_inside_image(const ellipse *e)
```

Comment:
Perhaps `ellipse_is_inside_image_area` ?
It isn't really checking the ellipse is inside a specific image, but rather that it is within the allowed area?

Reply:
Good point. Or maybe `ellipse_is_inside_bounds`.

Reply:
That would also be clearer.